File: typescript/src/testing-levels/unit-testing-with-mocks/checkout.ts
/* The solitarist setup: every collaborator is an interface, so every one of
   them can be replaced. https://endtoendtester.com/testing-levels/unit-testing-with-mocks */

export interface ChargeRequest {
  cents: number;
  reference: string;
  idempotencyKey: string;
}

export interface ChargeResult {
  ok: boolean;
  paymentIntentId?: string;
  reason?: string;
}

export interface PaymentGateway {
  charge(request: ChargeRequest): Promise<ChargeResult>;
}

export interface InventoryRepository {
  reserve(sku: string, quantity: number): Promise<boolean>;
  release(sku: string, quantity: number): Promise<void>;
}

export interface Clock {
  now(): number;
}

export interface Order {
  reference: string;
  sku: string;
  quantity: number;
  cents: number;
}

export type Outcome =
  | { ok: true; paymentIntentId: string }
  | { ok: false; reason: 'out-of-stock' | 'declined' };

export class Checkout {
  constructor(
    private readonly gateway: PaymentGateway,
    private readonly inventory: InventoryRepository,
    private readonly clock: Clock
  ) {}

  async place(order: Order): Promise<Outcome> {
    const reserved = await this.inventory.reserve(order.sku, order.quantity);
    if (!reserved) return { ok: false, reason: 'out-of-stock' };

    const charge = await this.gateway.charge({
      cents: order.cents,
      reference: order.reference,
      // Deterministic per order per day: a retry on the same day must not
      // charge twice, which is why the clock is injected rather than read.
      idempotencyKey: `${order.reference}-${new Date(this.clock.now()).toISOString().slice(0, 10)}`
    });

    if (!charge.ok) {
      // The compensating action. Nothing else in the system will do it.
      await this.inventory.release(order.sku, order.quantity);
      return { ok: false, reason: 'declined' };
    }

    return { ok: true, paymentIntentId: charge.paymentIntentId! };
  }
}

File: dotnet/Shop/Ordering/Checkout.cs
using Shop.Pricing;

namespace Shop.Ordering;

/// <summary>
/// Every dependency arrives through the constructor, including the clock —
/// which is what makes the idempotency key testable at all.
/// https://endtoendtester.com/practices/dependency-injection
/// </summary>
public sealed class Checkout(IPaymentGateway gateway, IInventory inventory, TimeProvider clock)
{
    private static readonly DiscountPolicy NoDiscount = new(new Money(int.MaxValue), 0);

    public async Task<Outcome> PlaceAsync(Order order, DiscountPolicy? policy = null, CancellationToken cancellationToken = default)
    {
        if (!await inventory.ReserveAsync(order.Sku, order.Quantity, cancellationToken))
        {
            return new Outcome.OutOfStock(order.Sku);
        }

        var priced = Pricing.Pricing.Apply(order.Amount, order.Tier, policy ?? NoDiscount);

        // Stable for the day, so a retry on the same day cannot charge twice.
        var key = $"{order.Reference}-{clock.GetUtcNow():yyyy-MM-dd}";
        var charge = await gateway.ChargeAsync(new ChargeRequest(priced.Total, order.Reference, key), cancellationToken);

        if (!charge.Succeeded)
        {
            // The compensating action. Nothing else in the system will do it.
            await inventory.ReleaseAsync(order.Sku, order.Quantity, cancellationToken);
            return new Outcome.Declined(charge.Reason ?? "declined");
        }

        return new Outcome.Placed(charge.PaymentIntentId!, priced.Total);
    }
}
